File: CSharp/ThrustRTCSharp/Context.cs
using System;
using ThrustRTCLR;
using System.Runtime.InteropServices;

namespace ThrustRTCSharp
{
    public struct dim_type
    {
        public uint x;
        public uint y;
        public uint z;

        public static implicit operator dim_type(uint x)
        {
            return new dim_type { x = x, y = 1, z = 1 };
        }
    };

    public partial class TRTC
    {
        static void set_libnvrtc_path(string path)
        {
            IntPtr p_path = Marshal.StringToHGlobalAnsi(path);
            Native.set_libnvrtc_path(p_path);
        }

        public static void Set_Verbose(bool verbose = true)
        {
            Native.set_verbose(verbose);
        }

        public static void Add_Include_Dir(string dir)
        {
            IntPtr p_dir = Marshal.StringToHGlobalAnsi(dir);
            Native.add_include_dir(p_dir);
        }

        public static void Add_Built_In_Header(string filename, string filecontent)
        {
            IntPtr p_filename = Marshal.StringToHGlobalAnsi(filename);
            IntPtr p_filecontent = Marshal.StringToHGlobalAnsi(filecontent);
            Native.add_built_in_header(p_filename, p_filecontent);
        }

        public static void Add_Inlcude_Filename(string filename)
        {
            IntPtr p_filename = Marshal.StringToHGlobalAnsi(filename);
            Native.add_include_filename(p_filename);
        }

        public static void Add_Code_Block(string code)
        {
            IntPtr p_code = Marshal.StringToHGlobalAnsi(code);
            Native.add_code_block(p_code);
        }

        public static void Wait()
        {
            Native.wait();
        }
    }

    partial class Internal
    {
        public static IntPtr[] ConvertStrList(string[] strs)
        {
            IntPtr[] p_strs = new IntPtr[strs.Length];
            for (int i = 0; i < strs.Length; i++)
                p_strs[i] = Marshal.StringToHGlobalAnsi(strs[i]);
            return p_strs;
        }

        public static IntPtr ConvertDVList(DeviceViewable[] args)
        {
            if (args.Length > 0)
            {
                IntPtr[] p_args = new IntPtr[args.Length];
                for (int i = 0; i < args.Length; i++)
                    p_args[i] = args[i].m_cptr;
                return Marshal.UnsafeAddrOfPinnedArrayElement(p_args, 0);
            }
            else
            {
                return IntPtr.Zero;
            }
        }

        public static dim_type_clr ConvertDimType(dim_type dim)
        {
            return new dim_type_clr { x = dim.x, y = dim.y, z = dim.z };
        }
    }

    public class Kernel
    {
        public readonly IntPtr m_cptr;

        public Kernel(string[] param_names, string body)
        {
            IntPtr[] p_param_names = Internal.ConvertStrList(param_names);
            IntPtr p_body = Marshal.StringToHGlobalAnsi(body);
            m_cptr = Native.kernel_create(p_param_names, p_body);
        }

        ~Kernel()
        {
            Native.kernel_destroy(m_cptr);
        }

        public int num_params()
        {
            return Native.kernel_num_params(m_cptr);
        }

        public int calc_optimal_block_size(DeviceViewable[] args, uint sharedMemBytes = 0)
        {
            IntPtr p_args = Internal.ConvertDVList(args);
            return Native.kernel_calc_optimal_block_size(m_cptr, p_args, sharedMemBytes);
        }

        public int calc_number_blocks(DeviceViewable[] args, int sizeBlock, uint sharedMemBytes = 0)
        {
            IntPtr p_args = Internal.ConvertDVList(args);
            return Native.kernel_calc_number_blocks(m_cptr, p_args, sizeBlock, sharedMemBytes);
        }

        public bool launch(dim_type gridDim, dim_type blockDim, DeviceViewable[] args, uint sharedMemBytes = 0)
        {
            IntPtr p_args = Internal.ConvertDVList(args);
            return Native.kernel_launch(m_cptr, Internal.ConvertDimType(gridDim), Internal.ConvertDimType(blockDim), p_args, sharedMemBytes);
        }
    }

    public class For
    {
        public readonly IntPtr m_cptr;

        public For(string[] param_names, string name_iter, string body)
        {
            IntPtr[] p_param_names = Internal.ConvertStrList(param_names);
            IntPtr p_name_iter = Marshal.StringToHGlobalAnsi(name_iter);
            IntPtr p_body = Marshal.StringToHGlobalAnsi(body);
            m_cptr = Native.for_create(p_param_names, p_name_iter, p_body);
        }
        
        ~For()
        {
            Native.for_destroy(m_cptr);
        }

        public int num_params()
        {
            return Native.for_num_params(m_cptr);
        }

        public bool launch(ulong begin, ulong end, DeviceViewable[] args)
        {
            IntPtr p_args = Internal.ConvertDVList(args);
            return Native.for_launch(m_cptr, begin, end, p_args);
        }

        public bool launch_n(ulong n, DeviceViewable[] args)
        {
            IntPtr p_args = Internal.ConvertDVList(args);
            return Native.for_launch_n(m_cptr, n, p_args);
        }
    }
}

File: CSharp/ThrustRTCLR/Context.cpp
#include "stdafx.h"
#include "ThrustRTCLR.h"
#include "TRTCContext.h"

namespace ThrustRTCLR 
{
	template<typename T>
	inline T* just_cast_it(IntPtr p)
	{
		return (T*)(void*)p;
	}

	void Native::set_libnvrtc_path(IntPtr p_path)
	{
		const char* c_path = just_cast_it<const char>(p_path);
		::set_libnvrtc_path(c_path);
	}

	void Native::set_verbose(bool verbose)
	{
		TRTC_Set_Verbose(verbose);
	}

	void Native::add_include_dir(IntPtr p_dir)
	{
		const char* c_dir = just_cast_it<const char>(p_dir);
		TRTC_Add_Include_Dir(c_dir);
	}

	void Native::add_built_in_header(IntPtr p_filename, IntPtr p_filecontent)
	{
		const char* c_filename = just_cast_it<const char>(p_filename);
		const char* c_filecontent = just_cast_it<const char>(p_filecontent);
		TRTC_Add_Built_In_Header(c_filename, c_filecontent);
	}

	void Native::add_include_filename(IntPtr p_filename)
	{
		const char* c_filename = just_cast_it<const char>(p_filename);
		TRTC_Add_Inlcude_Filename(c_filename);
	}

	void Native::add_code_block(IntPtr p_code)
	{
		const char* c_code = just_cast_it<const char>(p_code);
		TRTC_Add_Code_Block(c_code);
	}

	IntPtr Native::kernel_create(array<IntPtr>^ p_param_names, IntPtr p_code_body)
	{
		int num_params = p_param_names->Length;
		std::vector<const char*> params(num_params);
		for (int i = 0; i < num_params; i++)
			params[i] = just_cast_it<const char>(p_param_names[i]);
		const char* body = just_cast_it<const char>(p_code_body);
		TRTC_Kernel* cptr = new TRTC_Kernel(params, body);
		return IntPtr(cptr);
	}

	void Native::kernel_destroy(IntPtr p_kernel)
	{
		TRTC_Kernel* kernel = just_cast_it<TRTC_Kernel>(p_kernel);
		delete kernel;
	}

	int Native::kernel_num_params(IntPtr p_kernel)
	{
		TRTC_Kernel* kernel = just_cast_it<TRTC_Kernel>(p_kernel);
		return (int)kernel->num_params();
	}

	int Native::kernel_calc_optimal_block_size(IntPtr p_kernel, IntPtr p_args, uint32_t sharedMemBytes)
	{
		TRTC_Kernel* kernel = just_cast_it<TRTC_Kernel>(p_kernel);
		const DeviceViewable** args = just_cast_it<const DeviceViewable*>(p_args);
		int sizeBlock;
		if (kernel->calc_optimal_block_size(args, sizeBlock, sharedMemBytes))
			return sizeBlock;
		else
			return -1;
	}

	int Native::kernel_calc_number_blocks(IntPtr p_kernel, IntPtr p_args, int sizeBlock, uint32_t sharedMemBytes)
	{
		TRTC_Kernel* kernel = just_cast_it<TRTC_Kernel>(p_kernel);
		const DeviceViewable** args = just_cast_it<const DeviceViewable*>(p_args);
		int numBlocks;
		if (kernel->calc_number_blocks(args, sizeBlock, numBlocks, sharedMemBytes))
			return numBlocks;
		else
			return -1;
	}

	static dim_type to_cpp(dim_type_clr in)
	{
		return { in.x, in.y, in.z };
	}

	bool Native::kernel_launch(IntPtr p_kernel, dim_type_clr gridDim, dim_type_clr blockDim, IntPtr p_args, uint32_t sharedMemBytes)
	{
		TRTC_Kernel* kernel = just_cast_it<TRTC_Kernel>(p_kernel);
		const DeviceViewable** args = just_cast_it<const DeviceViewable*>(p_args);
		return kernel->launch(to_cpp(gridDim), to_cpp(blockDim), args, sharedMemBytes);
	}

	IntPtr Native::for_create(array<IntPtr>^ param_names, IntPtr p_name_iter, IntPtr p_code_body)
	{
		int num_params = param_names->Length;
		std::vector<const char*> params(num_params);
		for (int i = 0; i < num_params; i++)
			params[i] = just_cast_it<const char>(param_names[i]);
		const char* idx = just_cast_it<const char>(p_name_iter);
		const char* body = just_cast_it<const char>(p_code_body);
		TRTC_For* cptr = new TRTC_For(params, idx, body);
		return IntPtr(cptr);
	}

	void Native::for_destroy(IntPtr p_kernel)
	{
		TRTC_For* kernel = just_cast_it<TRTC_For>(p_kernel);
		delete kernel;
	}

	int Native::for_num_params(IntPtr p_kernel)
	{
		TRTC_For* kernel = just_cast_it<TRTC_For>(p_kernel);
		return (int)kernel->num_params();
	}

	bool Native::for_launch(IntPtr p_kernel, size_t begin, size_t end, IntPtr p_args)
	{
		TRTC_For* kernel = just_cast_it<TRTC_For>(p_kernel);
		const DeviceViewable** args = just_cast_it<const DeviceViewable*>(p_args);
		return kernel->launch(begin, end, args);
	}

}

